g:ale_lint_on_enter = 1 does not seem to work anymore (at least for Go)

I had to manually save the file to trigger the linting (which is probably the cmd shown at the bottom)...

---
 Current Filetype: go
Available Linters: ['go build', 'gofmt', 'golint', 'gometalinter', 'gosimple', 'go vet', 'staticcheck']
  Enabled Linters: ['gometalinter']
 Linter Variables:

let g:ale_go_gometalinter_options = ' --aggregate --fast --sort=line --vendor --vendored-linters --disable=gas --disable=goconst --disable=gocyclo '
 Global Variables:

let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {'go': ['gometalinter'], 'html': [], 'javascript': ['eslint']}
let g:ale_open_list = 0
let g:ale_set_highlights = 0
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 1
let g:ale_sign_error = '✖'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '⚠'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 0
  Command History:

(finished - exit code 1) ['/usr/local/bin/bash', '-c', 'gometalinter --include=''^/Users/svanharmelen/Documents/GoCode/src/github.com/hashicorp/packer/builder/cloudstack/step_create_instance.go.*$''  --aggregate --fast --sort=line --vendor --vendored-linters --disable=gas --disable=goconst --disable=gocyclo  ''/Users/svanharmelen/Documents/GoCode/src/github.com/hashicorp/packer/builder/cloudstack''']

<OUTPUT STARTS>
/Users/svanharmelen/Documents/GoCode/src/github.com/hashicorp/packer/builder/cloudstack/step_create_instance.go:68::warning: declaration of "err" shadows declaration at step_create_instance.go:51 (vetshadow)
<OUTPUT ENDS>
---

I think there is a problem with the regexp for include in gometalinter?

correct
```
$ gometalinter /tmp/ --include "/tmp/test.go.*$" 
../../../../../../../../../../../tmp/test.go:21:9:error: undeclared name: asfd (gotype)
```
incorrect
```
$ gometalinter /tmp/ --include "^/tmp/test.go.*$" 

```

As you can see if i add ^ the beginning like this plugin does it doesnt work. 

Thats because the output seems to always output relative paths and not absolute.

@jonaz check...

As for your issue, I don't think there is a problem with the regex used for `gometalinter`. I use this linter daily for months now without any issues and before that I used the `vim-go` version (which uses the exact same regex) for years.

So I guess you should include a way to reproduce this from within vim and add the output of `:ALEInfo` to show what command was exactly triggered.

See here an example of my output which is exactly what I would expect:

```
 Current Filetype: go
Available Linters: ['go build', 'gofmt', 'golint', 'gometalinter', 'gosimple', 'go vet', 'staticcheck']
  Enabled Linters: ['gometalinter']
 Linter Variables:

let g:ale_go_gometalinter_options = ' --aggregate --fast --sort=line --vendor --vendored-linters --disable=gas --disable=goconst --disable=gocyclo '
 Global Variables:

let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {'go': ['gometalinter'], 'html': [], 'javascript': ['eslint']}
let g:ale_open_list = 0
let g:ale_set_highlights = 0
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 1
let g:ale_sign_error = '✖'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '⚠'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 0
  Command History:

(finished - exit code 1) ['/usr/local/bin/bash', '-c', 'gometalinter --include=/Users/svanharmelen/Documents/GoCode/src/demo/main.go  --aggregate --fast --sort=line --vendor --vendored-linters --disable=gas --disable=goconst --disable=gocyclo  ''/Users/svanharmelen/Documents/GoCode/src/demo''']

<OUTPUT STARTS>
/Users/svanharmelen/Documents/GoCode/src/demo/main.go:5:2:error: bla (invalid operand) is not used (gotype)
/Users/svanharmelen/Documents/GoCode/src/demo/main.go:5:2:error: undeclared name: bla (gotype)
/Users/svanharmelen/Documents/GoCode/src/demo/main.go:8:1:warning: exported function Bla should have comment or be unexported (golint)
<OUTPUT ENDS>
```

--- a/autoload/ale.vim
+++ b/autoload/ale.vim
@@ -52,7 +52,7 @@ function! ale#ShouldDoNothing(buffer) abort
     endif
 
     " Do nothing for blacklisted files
-    if index(g:ale_filetype_blacklist, &filetype) >= 0
+    if index(g:ale_filetype_blacklist, getbufvar(a:buffer, '&filetype')) >= 0
         return 1
     endif
 
@@ -118,7 +118,7 @@ function! s:ALEQueueImpl(delay, linting_flag, buffer) abort
     " Remember that we want to check files for this buffer.
     " We will remember this until we finally run the linters, via any event.
     if a:linting_flag is# 'lint_file'
-        let s:should_lint_file_for_buffer[bufnr('%')] = 1
+        let s:should_lint_file_for_buffer[a:buffer] = 1
     endif
 
     if s:lint_timer != -1

